libraries/node-core-library/src/Import.ts

```diff
@@ -83,4 +173,109 @@ export class Import {
+   */
+  public static resolveModule(options: IImportResolveModuleOptions): string {
+    const { modulePath } = options;
+
+    if (path.isAbsolute(modulePath)) {
+      return modulePath;
+    }
+
+    const normalizedRootPath: string = FileSystem.getRealPath(options.baseFolderPath);
+
+    if (modulePath.startsWith('.')) {
+      // This looks like a conventional relative path
+      return path.resolve(normalizedRootPath, modulePath);
+    }
+
+    if (options.includeSystemModules === true && Import._builtInModules.has(modulePath)) {
+      return modulePath;
+    }
+
+    if (options.allowSelfReference === true) {
+      const ownPackage: IPackageDescriptor | undefined = Import._getPackageName(options.baseFolderPath);
+      if (ownPackage && modulePath.startsWith(ownPackage.packageName)) {
+        const packagePath: string = modulePath.substr(ownPackage.packageName.length + 1);
+        return path.resolve(ownPackage.packageRootPath, packagePath);
+      }
+    }
+
+    try {
+      return Resolve.sync(
+        // Append a slash to the package name to ensure `resolve.sync` doesn't attempt to return a system package
+        options.includeSystemModules !== true && modulePath.indexOf('/') === -1
+          ? `${modulePath}/`
+          : modulePath,
+        {
+          basedir: normalizedRootPath,
+          preserveSymlinks: false
+        }
+      );
+    } catch (e) {
+      throw new Error(`Cannot find module "${modulePath}" from "${options.baseFolderPath}".`);
+    }
+  }
+
+  /**
+   * Resolve the path to a package, from the perspective of a specified folder.
+   */
+  public static resolvePackage(options: IImportResolvePackageOptions): string {
+    const { packageName } = options;
+
+    if (options.includeSystemModules && Import._builtInModules.has(packageName)) {
+      return packageName;
+    }
+
+    const normalizedRootPath: string = FileSystem.getRealPath(options.baseFolderPath);
+
+    if (options.allowSelfReference) {
+      const ownPackage: IPackageDescriptor | undefined = Import._getPackageName(options.baseFolderPath);
+      if (ownPackage && ownPackage.packageName === packageName) {
+        return ownPackage.packageRootPath;
+      }
+    }
+
+    try {
+      const resolvedPath: string = Resolve.sync(packageName, {
+        basedir: normalizedRootPath,
+        preserveSymlinks: false,
+        packageFilter: (pkg: { main: string }): { main: string } => {
+          // In case the "main" property isn't defined, set it to something we know will exist
+          pkg.main = 'package.json';
+          return pkg;
+        }
+      });
+
+      const packagePath: string = path.dirname(resolvedPath);
+      const packageJson: IPackageJson = PackageJsonLookup.instance.loadPackageJson(
+        path.join(packagePath, 'package.json')
+      );
+      if (packageJson.name === packageName) {
+        return packagePath;
+      } else {
+        throw new Error();
+      }
```

Comment:
Seems like we can delete this code. Since we hardwired `main` to be package.json, that file must exist, otherwise `resolve.sync()` would have failed.

Reply:
Also I can't think of any realistic case where `packageJson.name !== packageName` could happen.

Reply:
The

```
      if (packageJson.name === packageName) {
        return packagePath;
      } else {
        throw new Error();
      }
```

code ensures that what was passed to this function is a package name and not a module path.

Reply:
@iclanton It might be better to use the `PackageName` API for that, since it is a cheap syntax test that doesn't need to parse the file from disk. And it will give better errors.